api/scanner/media_type/magic.go

```diff
@@ -0,0 +1,73 @@
+package media_type
+
+import (
+	"strings"
+	"sync"
+
+	magic "github.com/hosom/gomagic"
+	"github.com/photoview/photoview/api/log"
+)
+
+var libmagic struct {
+	mu       sync.Mutex
+	libmagic *libMagic
+	err      error
+}
+
+func init() {
+	libmagic.libmagic, libmagic.err = newLibMagic()
+	if libmagic.err != nil {
+		libmagic.libmagic = nil
+		log.Error("Init libmagic error.", "error", libmagic.err)
+	}
+}
+
+// GetMediaType returns a media type of file `f`.
+// This function is thread-safe.
+func GetMediaType(f string) MediaType {
+	libmagic.mu.Lock()
```

Comment:
`libmagic` uses a table to check the file magic number, which is not a thread-safe action.

Reply:
Does it mean that 2 parallel calls might/will interfere with each other even if they process different images?

With the current implementation, does it mean that only 1 image at a time can be processed or do I misunderstand it? If yes, is there a way to overcome this restriction e.g. by running parallel probing not in threads, but in sub-processes to isolate these instances of the table?
My concern here is that running the image probing in a single thread might slow down the scanning job of a large media gallery with hundreds of thousands of images (even while probing a single image is a fast operation)

Reply:
> Does it mean that 2 parallel calls might/will interfere with each other even if they process different images?

Yes.

 I don't think it'll slow parallel jobs. This function call returns very quick as it just reads a small number bytes of a file, but not the whole file. `magic.Open()` is a heavier call than `magic.File()`, because it not only opens files from the disk, but also allocs a table in memory and parse those files into the memory.

I don't see it's a bottle-neck of the performance. So keep it simple.

Reply:
> I don't think it'll slow parallel jobs

Could you please explain what do you mean by parallel jobs? I guess, this might be the root cause of my misunderstanding...
In my understanding, we run a single scan job for the media collection. Then it catalogues all the sub-folders and media files inside, and runs a separate scanning task per file in parallel. 
As far as I understand, these parallel tasks are running in threads and after this locking is merged, there will be only 1 thread at least for probing all the images in the collection. So, if we have hundreds of thousands of images, just iterating over them and reading a few KB of data would take noticeably longer time than doing this in parallel.

Please let me know where am I wrong and why do you see no significant difference in the scanning performance.

Reply:
Locking is not a problem with parallel tasks, performance is.

I added [benchmarks](https://github.com/photoview/photoview/actions/runs/12223889525/job/34095964716?pr=1124#step:6:73) for two functions:

```
BenchmarkMagicAll-4    	    3501	    335427 ns/op	      40 B/op	       2 allocs/op
BenchmarkMagicType-4   	    8960	    125508 ns/op	      32 B/op	       1 allocs/op
```

`MagicAll` calls `Open`/`File`/`Close`, and `MagicType` calls `File` with locker. You can see that `MagicType` is just half the calling time of `MagicAll`, with less memory requirement.

The time-consuming operation is converting images. Even with the lock during `GetMediaType()`, it won't block conversions in parallel.

Reply:
Thanks for the detailed explanation!

Reply:
Another way is that we could pre-fork goroutines for running jobs, and each goroutine holds a libmagic instance. It requires changing the current code largely. Maybe we could do that later if necessary.

As current code, it's hard to let each goroutine/thread hold a libmagic. As the function is not heavy (compared to convert files), locking is easy and enough for performance

Reply:
Sure, let's go with the current solution and see if there is a bottleneck (or close to it) in practice on large media library setups.
It would be great if we could log some performance indicators and analyze them later from the data shared by users as part of performance-related tickets. This would give us an understanding of bottlenecks in the job.